beater/java_attacher/java_attacher.go

```diff
@@ -25,50 +25,369 @@ import (
+	for _, jvm := range jvms {
+		j.logger.Debugf("PID: %v, version: %v, start-time: %v, user: '%v', command: '%v'",
+			jvm.pid, jvm.version, jvm.startTime, jvm.user, jvm.command)
+	}
+}
 
-func (j JavaAttacher) Run(ctx context.Context) error {
-	cmd := j.build(ctx)
-	j.logger.Debugf("starting java attacher with command: %s", strings.Join(cmd.Args, " "))
+// this function blocks until discovery has ended, an error occurred or the context had been cancelled
+func (j *JavaAttacher) discoverAllRunningJavaProcesses(ctx context.Context) (map[string]*JvmDetails, error) {
+	jvms := make(map[string]*JvmDetails)
+
+	// We can't discover start time at this point because both `start` and `lstart` don't follow a strict-enough format,
+	// which may interfere with output parsing.
+	cmd := exec.CommandContext(ctx, "ps", "-A", "-ww", "-o", "user=,uid=,gid=,pid=,comm=")
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, err
+	}
+	if err := cmd.Start(); err != nil {
+		return nil, err
+	}
+	scanner := bufio.NewScanner(stdout)
+	for scanner.Scan() {
+		line := scanner.Text()
+		psOutputLineParts := strings.Fields(line)
+		// we cannot rely on exact number of parts because the command output may contain spaces, however, since it is printed last,
+		// we are ok with just taking the command's first part in the case of Java processes
+		if len(psOutputLineParts) > 4 {
+			command := psOutputLineParts[4]
+			if strings.Contains(strings.ToLower(command), "java") {
+				pid := psOutputLineParts[3]
+				jvms[pid] = &JvmDetails{
+					psOutputLineParts[0],
+					psOutputLineParts[1],
+					psOutputLineParts[2],
+					pid,
+					"unknown",
+					command,
+					"unknown",
+					"",
+				}
+			}
+		} else {
+			j.logger.Errorf("Unexpected number of output fields (%v) from command '%v'", len(psOutputLineParts), strings.Join(cmd.Args, " "))
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		j.logger.Errorf("error reading output for command '%v': %v", strings.Join(cmd.Args, " "), err)
+	}
+	err = cmd.Wait()
+	if err != nil {
+		j.logger.Errorf("error executing command '%v': %v", strings.Join(cmd.Args, " "), err)
+	}
+	return jvms, nil
+}
+
+// this function blocks until the first of the following occurs:
+// 	1.	the requested operation was executed and terminated (orderly or erroneously) for all received JVMs
+//	2.	the joint execution timed out
+// 	3.	the context either had been cancelled
+func (j *JavaAttacher) executeForEachJvm(ctx context.Context, jvms map[string]*JvmDetails,
+	executable func(ctx context.Context, jvm *JvmDetails) error, removeOnError bool, timeout time.Duration) error {
+	if len(jvms) == 0 {
+		return nil
+	}
+	timeoutCtx, cancelFunc := context.WithTimeout(ctx, timeout)
+	defer cancelFunc()
+	var wg sync.WaitGroup
+	for pid, jvm := range jvms {
+		wg.Add(1)
+		go func(jvm *JvmDetails, pid string) {
+			err := executable(timeoutCtx, jvm)
+			if err != nil {
+				j.logger.Debug(err)
+				if removeOnError {
+					delete(jvms, pid)
```

Comment:
As this is in a goroutine, there may be concurrent map deletions. We need to synchronise these.

Reply:
Ohh man, that's right! Coming from Java, I looked whether it is even safe to remove from a map while iterating, but neglected this aspect when changed from sequential to concurrent execution. Thanks for catching this!